src/Cli/dotnet/commands/dotnet-test/MSBuildUtility.cs

```diff
@@ -0,0 +1,218 @@
+
+            // Return the resolved solution file path
+            return solutionFullPath;
+        }
+
+        internal static bool IsBinaryLoggerEnabled(List<string> args, out string binLogFileName)
+        {
+            binLogFileName = string.Empty;
+            var binLogArgs = new List<string>();
+
+            foreach (var arg in args)
+            {
+                if (arg.StartsWith("/bl:") || arg.Equals("/bl")
+                    || arg.StartsWith("--binaryLogger:") || arg.Equals("--binaryLogger")
+                    || arg.StartsWith("-bl:") || arg.Equals("-bl"))
+                {
+                    binLogArgs.Add(arg);
+
+                }
+            }
+
+            if (binLogArgs.Count > 0)
+            {
+                // Remove all BinLog args from the list of args
+                args.RemoveAll(arg => binLogArgs.Contains(arg));
+
+                // Get BinLog filename
+                var binLogArg = binLogArgs.LastOrDefault();
+
+                if (binLogArg.Contains(CliConstants.Colon))
+                {
+                    var parts = binLogArg.Split(CliConstants.Colon, 2);
+                    binLogFileName = !string.IsNullOrEmpty(parts[1]) ? parts[1] : CliConstants.BinLogFileName;
+                }
+                else
+                {
+                    binLogFileName = CliConstants.BinLogFileName;
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        private static BuildParameters GetBuildParameters(ProjectCollection projectCollection, BuildOptions buildOptions)
+        {
+            BuildParameters parameters = new(projectCollection)
+            {
+                Loggers = [new ConsoleLogger(LoggerVerbosity.Quiet)]
+            };
+
+            if (!buildOptions.AllowBinLog)
+                return parameters;
+
+            parameters.Loggers =
+            [
+                .. parameters.Loggers,
+                .. new[]
+                {
+                    new BinaryLogger
+                    {
+                        Parameters = buildOptions.BinLogFileName
+                    }
+                },
+            ];
+
+            return parameters;
+        }
+
+        private static Dictionary<string, string> GetGlobalProperties(BuildOptions buildOptions)
+        {
+            var globalProperties = new Dictionary<string, string>();
+
+            if (!string.IsNullOrEmpty(buildOptions.Configuration))
+            {
+                globalProperties[CliConstants.Configuration] = buildOptions.Configuration;
+            }
+
+            if (!string.IsNullOrEmpty(buildOptions.RuntimeIdentifier))
+            {
+                globalProperties[CliConstants.RuntimeIdentifier] = buildOptions.RuntimeIdentifier;
+            }
+
+            return globalProperties;
+        }
+
+        private static string[] GetCommands(bool hasNoRestore, bool hasNoBuild)
```

Comment:
Do you need to return as array? This is creating 2 collections for not so much. If you care about immutability then ImmutableArray is better and has a builder.